association.go

```diff
@@ -924,24 +1280,136 @@ func (a *Association) resetStreams() *packet {
 	}})
 }
 
-// sendPayloadData sends the data chunks.
-func (a *Association) sendPayloadData(chunks []*chunkPayloadData) error {
-	a.lock.Lock()
+func (a *Association) peekNextPendingData() *chunkPayloadData {
+	return a.pendingQueue.peek()
+}
+
+// Move the chunk peeked with peekNextPendingData() to the inflightQueue.
+func (a *Association) movePendingDataChunkToInflightQueue(c *chunkPayloadData) {
+	if err := a.pendingQueue.pop(c); err != nil {
+		a.log.Errorf("failed to pop from pending queue: %s", err.Error())
+	}
+
+	// Assign TSN
+	c.tsn = a.generateNextTSN()
+
+	c.since = time.Now() // use to calculate RTT and also for maxPacketLifeTime
+	c.nSent = 1          // being sent for the first time
+
+	a.checkPartialReliabilityStatus(c)
+
+	a.log.Tracef("sending tsn=%d ssn=%d sent=%d", c.tsn, c.streamSequenceNumber, c.nSent)
+
+	// Push it into the inflightQueue
+	a.inflightQueue.pushNoCheck(c)
+}
+
+// popPendingDataChunksToSend pops chunks from the pending queues as many as
+// the cwnd and rwnd allows to send.
+// The caller should hold the lock.
+func (a *Association) popPendingDataChunksToSend() []*chunkPayloadData {
+	chunks := []*chunkPayloadData{}
+
+	if a.pendingQueue.size() > 0 {
```

Comment:
We check this several times, should we just make an empty()?

Reply:
size() is used in the tests (payload_queue_test.go) a lot. I think adding empty() would also check the length of the internal queue (no performance gain), and size() has more "information" than empty().  Let me keep those lines...